src/frontends/pytorch/src/transforms/prim_tuple_unpack_parameter_replacer.cpp

```diff
@@ -0,0 +1,97 @@
+        parameters.pop();
+        auto consumers = parameter->get_output_target_inputs(0);
+        size_t num_outputs = 0;  // number of outputs in each unpack consumer should match
+        bool all_unpacks = true;
+
+        // collects all outputs per each consumer operation for this tuple Parameter
+        std::vector<OutputVector> consumer_outputs;
+
+        for (const auto& consumer : consumers) {
+            auto node = consumer.get_node()->shared_from_this();
+            auto tuple_unpack = cast_fw_node(node, "prim::TupleUnpack");
+            if (!tuple_unpack) {
+                all_unpacks = false;
+                break;
+            }
+            if (num_outputs == 0) {
+                num_outputs = node->get_output_size();
+            } else if (num_outputs != node->get_output_size()) {
+                std::cerr << "[ PT FE WARNING ] Unpack node " << node
+                          << " as one of the consumers of tuple object has number of outputs "
+                          << node->get_output_size() << " not matching number of outputs " << num_outputs
+                          << " for other consumer.\n";
+                all_unpacks = false;
+                break;
+            }
+            consumer_outputs.push_back(node->outputs());
+        }
+
+        if (!all_unpacks || consumer_outputs.empty()) {
+            // if at least one consumer is not an unpack-like op or there are not matching number of unpacked objects,
+            // we cannot replace other unpacks even if they exist, leaving Unpack-op(s) in the graph for this Parameter
+            continue;
+        }
+
+        for (size_t i = 0; i < num_outputs; ++i) {
+            // Merged partial shape and element type among all the consumers of i-th result of unpack ops
+            PartialShape ps = PartialShape::dynamic();
+            element::Type et = element::dynamic;
+            std::set<Input<Node>> inputs;
+
+            for (const auto& outputs : consumer_outputs) {
+                auto output = outputs[i];
+                OPENVINO_ASSERT(PartialShape::merge_into(ps, output.get_partial_shape()),
+                                "Consumers for unpack op have incompatible shape");
+                OPENVINO_ASSERT(element::Type::merge(et, et, output.get_element_type()),
+                                "Consumers for unpack op have incompatible types");
+                auto target_inputs = output.get_target_inputs();
+                inputs.insert(target_inputs.begin(), target_inputs.end());
+            }
+
+            auto new_parameter = std::make_shared<ov::op::v0::Parameter>(et, ps);
+
+            for (auto input : inputs) {
+                auto names = input.get_tensor().get_names();
+                input.replace_source_output(new_parameter->output(0));
+                new_parameter->output(0).add_names(names);
+            }
+
+            // TODO: Assign correct names
+            model->add_parameters({new_parameter});
```

Comment:
I don't think this will preserve order of inputs for the case when tuple is followed by a non tuple input.

Reply:
Real thing! Thanks for catching it. This is copied-and-pasted from a similar `Result`s flattening transformation at src/frontends/pytorch/src/transforms/prim_list_tuple_construct_replacer.cpp. Just added a couple of tests and the problem is really reproduced for both transformations. Will fix both.